[JSC] Tier up when liveness / fullness profile rate does not change

Constellation opened the pull request:
#### 67685bf451f9053a95ae95b3122e368b3f6e4f33
<pre>
[JSC] Tier up when liveness / fullness profile rate does not change
<a href="https://bugs.webkit.org/show_bug.cgi?id=296607">https://bugs.webkit.org/show_bug.cgi?id=296607</a>
<a href="https://rdar.apple.com/156976025">rdar://156976025</a>

Reviewed by Mark Lam and Keith Miller.

If a CodeBlock size is large enough, then there is a chance that some
part of code is just a dead code. This can happen in a framework code
since it is crafted as a generic function. In this case, we track
whether the liveness / fullness rate changes from the previous sampling
time, and if it is not changed, we say that we are reaching to the
plateau and we do not expect that we will get more information with
retrying. Thus we will start compiling it in DFG. This patch records
the previous liveness / fullness rate and detect it is not changed.
And then we tier up early based on that information.

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::shouldOptimizeNowFromBaseline):
* Source/JavaScriptCore/jit/BaselineJITCode.h:
(JSC::BaselineJITCode::livenessRate const):
(JSC::BaselineJITCode::setLivenessRate):
(JSC::BaselineJITCode::fullnessRate const):
(JSC::BaselineJITCode::setFullnessRate):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: <a href="https://commits.webkit.org/298008@main">https://commits.webkit.org/298008@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f96c36537cf8f332fa867aa27466cc167a263786

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113868 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33620 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24079 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120031 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/64657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f8713b9a-30f9-417b-b896-72f14e0ae2ac) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34248 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42181 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/86564 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/64657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d953e7ab-4f41-4226-94ec-4e2521bb7cba) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116816 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27255 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102276 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66937 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26463 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20406 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63753 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/106317 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96622 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20522 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123265 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/112431 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40915 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30458 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41287 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98488 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95165 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40294 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18098 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18263 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40783 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46294 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/136637 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40425 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36594 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43724 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42183 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian reviewed (approved):
r=me

kmiller68 commented on Source/JavaScriptCore/jit/BaselineJITCode.h:
```diff
@@ -94,13 +94,22 @@ class BaselineJITCode : public DirectJITCode, public MathICHolder {
     std::unique_ptr<PCToCodeOriginMap> m_pcToCodeOriginMap;
+private:
+    double m_livenessRate { 0 };
+    double m_fullnessRate { 0 };
```
Nit: probably worth a comment here saying what these are. e.g. something like:

```
     // The percentage of ValueProfiles that had some profiling data in them.
     double m_livenessRate { 0 };
     // The percentage of ValueProfile buckets that had a value in them.
     double m_fullnessRate { 0 };
```

Maybe we should also make it clear this is about ValueProfiles e.g. `m_valueProfileFullnessRate`

Constellation replied:
Sounds good, added.

kmiller68 reviewed (approved):
r=me too with nit

kmiller68 commented on Source/JavaScriptCore/bytecode/CodeBlock.cpp:
```diff
@@ -3070,6 +3071,29 @@ bool CodeBlock::shouldOptimizeNowFromBaseline()
+        double previousLivenessRate = baselineJITCode->livenessRate();
+        double previousFullnessRate = baselineJITCode->fullnessRate();
+        if (static_cast<unsigned>(m_optimizationDelayCounter) + 1 >= Options::minimumOptimizationDelay()) {
+            if (static_cast<int32_t>(this->bytecodeCost()) >= 5000) {
```
Nit: Maybe this should be an `Option`?

Constellation replied:
Changed.

Constellation committed:
[JSC] Tier up when liveness / fullness profile rate does not change
https://bugs.webkit.org/show_bug.cgi?id=296607
rdar://156976025

Reviewed by Mark Lam and Keith Miller.

If a CodeBlock size is large enough, then there is a chance that some
part of code is just a dead code. This can happen in a framework code
since it is crafted as a generic function. In this case, we track
whether the liveness / fullness rate changes from the previous sampling
time, and if it is not changed, we say that we are reaching to the
plateau and we do not expect that we will get more information with
retrying. Thus we will start compiling it in DFG. This patch records
the previous liveness / fullness rate and detect it is not changed.
And then we tier up early based on that information.

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::shouldOptimizeNowFromBaseline):
* Source/JavaScriptCore/jit/BaselineJITCode.h:
(JSC::BaselineJITCode::livenessRate const):
(JSC::BaselineJITCode::setLivenessRate):
(JSC::BaselineJITCode::fullnessRate const):
(JSC::BaselineJITCode::setFullnessRate):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: https://commits.webkit.org/298008@main
